### PR TITLE
KAFKA-13403 Fix KafkaServer crashes when deleting topics due to the race in log deletion

### PR DESCRIPTION
**Pull request by arunmathew88**
Update the walkFileTree override implementation to handle parallel file deletion. So as to prevent crashing of Kafka broker process itself when logs are deleted by other threads due to retention expiry.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by arunmathew88** [comment]
retest this please

File: clients/src/main/java/org/apache/kafka/common/utils/Utils.java [hunk 1]
```diff
@@ -853,15 +853,21 @@ public static void delete(final File rootFile) throws IOException {
...
             public FileVisitResult visitFileFailed(Path path, IOException exc) throws IOException {
-                // If the root path did not exist, ignore the error; otherwise throw it.
-                if (exc instanceof NoSuchFileException && path.toFile().equals(rootFile))
-                    return FileVisitResult.TERMINATE;
+                if (exc instanceof NoSuchFileException) {
+                    if (path.toFile().equals(rootFile)) {
+                        // If the root path did not exist, ignore the error and terminate;
+                        return FileVisitResult.TERMINATE;
+                    } else {
+                        //  Otherwise, just continue walking because we don't have to delete this file.
```
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/common/utils/Utils.java]: nit: additional space after //
Also, could we mention that the file might already be deleted by other threads?

**Review by showuon — COMMENTED** [review]
@arunmathew88 , thanks for the PR. The change LGTM! Could we add a test under `UtilsTest#testRecursiveDelete`? I'm thinking we can use Mockito to do some "inexisted" file tests. WDYT?

**Comment by divijvaidya** [comment]
Hey @arunmathew88 
I know it has been a while but I was wondering if you would like to update this PR again? We have multiple bugs which are fixed by it.

**Comment by arunmathew88** [comment]
> Hey @arunmathew88 I know it has been a while but I was wondering if you would like to update this PR again? We have multiple bugs which are fixed by it.

Hello @divijvaidya , Thanks for the nudge. Was working on a test case as requested in review comments.
Have coded the changes, will push them in some time.

**Comment by divijvaidya** [comment]
Hey @arunmathew88 
One concern with the current change is that there may be places in the code base which rely on Util.delete to throw a fileNotFound exception. Have you verified that there are no such usages of this function? If we are not sure, then a safer path is to add a new function Utils.deleteIfExists and use that instead.

cc: @ocadaruma

**Comment by ocadaruma** [comment]
As of current trunk (0017bb3db5476479a9cfde2ec667949b6b001451), there are only 3 places which calls `Utils.delete` (excluding tests)
- `UnifiedLog.delete` which expects all children have been removed before calling `Utils.delete` (unless race happens like this PR addresses) so we don't have a motivation to throw NoSuchFileException
- `KafkaCSVMetricsReporter.init` Honestly I'm not familiar with this class but seems it is to clean-up the output dir before starting, so not throwing NoSuchFileException looks safe (in the first place the race wouldn't happen here I guess)
- `CoreUtils.delete` Actually this is used only for testing

So I guess it's safe to stop throwing NoSuchFileException at all. (Since `Utils.delete` is supposed to delete all files recursively, I can't imagine a case which throwing the exception is worth because it just implies the file is already deleted so no need to take any action)

**Comment by divijvaidya** [comment]
Thank you for the analysis @ocadaruma. In that case using deleteIfExists makes sense to me. Let's wait for @arunmathew88 to rebase from trunk and update this PR

**Comment by arunmathew88** [comment]
@divijvaidya I have pushed the test code as well. Please review.

**Comment by arunmathew88** [comment]
retest this please

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the fix.

**Comment by arunmathew88** [comment]
Thanks @showuon .
@divijvaidya What are the further steps needed to get this merged?

**Comment by divijvaidya** [comment]
@arunmathew88, the ball is in my court right now. Please give me a few days to review this. There are other places apart from what @ocadaruma mentioned such as `StateDirectory.java` which utilize this function. I want to ensure that this change in behaviour does not impact the logic at those places.

**Comment by arunmathew88** [comment]
@divijvaidya If it helps, we have been running this patch in our primary cluster which has a lot of topic delete-recreates and has not had a related issue for almost 2 years.

**Comment by arunmathew88** [comment]
@divijvaidya A gentle nudge.

**Comment by arunmathew88** [comment]
@showuon @divijvaidya Please take a look into this, as it is long pending.
Thanks.

**Review by divijvaidya — APPROVED** [review]
Apologies for the long delay Arun! I have looked at all places in the code where this utility is used. 

One of the fishy places was KafkaStreams.cleanup() where user behaviour could be impacted. Prior to this change, they will receive an exception if we try to delete state directory and it doesn't exist but after this they wouldn't. However, our intended behaviour is not to throw an exception if `stateDir` doesn't exist which is highlighted by the check here:
https://github.com/apache/kafka/blob/b8ad609541bb8b224d16ce832a08f51017002bc1/streams/src/main/java/org/apache/kafka/streams/processor/internals/StateDirectory.java#L409 
Hence, this change should be ok.

Thanks for your patience all this time again. Could you please rebase again one last time with trunk so that we can run this against the latest set of tests?

**Comment by arunmathew88** [comment]
@divijvaidya Thanks for the approval and update.
I have just pushed rebased commits.

**Comment by divijvaidya** [comment]
We have a test failing for this PR which has not failed in trunk for last 28 days [2]. @arunmathew88 can you please debug why this PR is causing that test to fail.

[1] https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-11438/6/
[2] https://ge.apache.org/scans/tests?search.relativeStartTime=P28D&search.rootProjectNames=kafka&search.tags=trunk&search.timeZoneId=Europe%2FBerlin&tests.container=kafka.log.remote.RemoteIndexCacheTest&tests.test=testIndexFileAlreadyExistOnDiskButNotInCache()

**Comment by arunmathew88** [comment]
@divijvaidya The test seems to be flaky, and not due to this PR.

**Comment by ocadaruma** [comment]
+1 with @arunmathew88. The failure doesn't seem to relate to this PR.

reasons:
- `testIndexFileAlreadyExistOnDiskButNotInCache()` seems to be failing sometimes even for recently
- The only place `testIndexFileAlreadyExistOnDiskButNotInCache()` calls `Utils.delete` (which potentially related to this PR) is at `cleanUp`, but the stacktrace in [2] shows the test fails in the middle of the test, not in `cleanUp`.

@divijvaidya Could you take a look?

**Comment by soarez** [comment]
@arunmathew88 could you rebase this to resolve the conflicts?

**Review by soarez — APPROVED** [review]
@arunmathew88: Thanks for the patch, for rebasing, and for all your patience through this.

Triggered a second CI, as the pipeline for JDK 21 did not finish on the first one:

* https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-11438/8/pipeline/
* https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-11438/9/pipeline/

No related tests failing. LGTM

**Pull request merged**